File: uav-weather-collector/src/main/java/com/uav/weather/model/WeatherData.java
package com.uav.weather.model;

import java.util.HashMap;
import java.util.Map;

public class WeatherData {
    private String source;
    private String droneId;
    private long timestamp;
    private double latitude, longitude, altitude;
    private double temperature, humidity, windSpeed, windDirection, windGust;
    private double pressure, visibility, turbulence, precipitation;
    private double dataQuality = 1.0;

    public Map<String, Object> toMap() {
        Map<String, Object> map = new HashMap<>();
        map.put("source", source);
        map.put("drone_id", droneId);
        map.put("timestamp", timestamp);
        map.put("latitude", latitude);
        map.put("longitude", longitude);
        map.put("altitude", altitude);
        map.put("temperature", temperature);
        map.put("humidity", humidity);
        map.put("wind_speed", windSpeed);
        map.put("wind_direction", windDirection);
        map.put("wind_gust", windGust);
        map.put("pressure", pressure);
        map.put("visibility", visibility);
        map.put("turbulence", turbulence);
        map.put("precipitation", precipitation);
        map.put("data_quality", dataQuality);
        return map;
    }

    public String getSource() { return source; }
    public void setSource(String source) { this.source = source; }
    public String getDroneId() { return droneId; }
    public void setDroneId(String droneId) { this.droneId = droneId; }
    public long getTimestamp() { return timestamp; }
    public void setTimestamp(long timestamp) { this.timestamp = timestamp; }
    public double getLatitude() { return latitude; }
    public void setLatitude(double latitude) { this.latitude = latitude; }
    public double getLongitude() { return longitude; }
    public void setLongitude(double longitude) { this.longitude = longitude; }
    public double getAltitude() { return altitude; }
    public void setAltitude(double altitude) { this.altitude = altitude; }
    public double getTemperature() { return temperature; }
    public void setTemperature(double temperature) { this.temperature = temperature; }
    public double getHumidity() { return humidity; }
    public void setHumidity(double humidity) { this.humidity = humidity; }
    public double getWindSpeed() { return windSpeed; }
    public void setWindSpeed(double windSpeed) { this.windSpeed = windSpeed; }
    public double getWindDirection() { return windDirection; }
    public void setWindDirection(double windDirection) { this.windDirection = windDirection; }
    public double getWindGust() { return windGust; }
    public void setWindGust(double windGust) { this.windGust = windGust; }
    public double getPressure() { return pressure; }
    public void setPressure(double pressure) { this.pressure = pressure; }
    public double getVisibility() { return visibility; }
    public void setVisibility(double visibility) { this.visibility = visibility; }
    public double getTurbulence() { return turbulence; }
    public void setTurbulence(double turbulence) { this.turbulence = turbulence; }
    public double getPrecipitation() { return precipitation; }
    public void setPrecipitation(double precipitation) { this.precipitation = precipitation; }
    public double getDataQuality() { return dataQuality; }
    public void setDataQuality(double dataQuality) { this.dataQuality = dataQuality; }
}

File: mobile/lib/models/weather_data.dart
class WeatherDataModel {
  const WeatherDataModel({
    this.source = '',
    this.droneId,
    required this.timestamp,
    this.latitude = 0,
    this.longitude = 0,
    this.altitude,
    this.temperature,
    this.humidity,
    this.windSpeed,
    this.windDirection,
    this.windGust,
    this.pressure,
    this.visibility,
    this.turbulence,
    this.precipitation,
    this.dataQuality = 1.0,
    this.weatherCondition,
    this.extra,
  });

  factory WeatherDataModel.fromJson(Map<String, dynamic> json) {
    return WeatherDataModel(
      source: json['source'] as String? ?? '',
      droneId: json['drone_id']?.toString(),
      timestamp: _parseTimestamp(json['timestamp']),
      latitude: (json['latitude'] ?? 0).toDouble(),
      longitude: (json['longitude'] ?? 0).toDouble(),
      altitude: json['altitude']?.toDouble(),
      temperature: json['temperature']?.toDouble(),
      humidity: json['humidity']?.toDouble(),
      windSpeed: json['wind_speed']?.toDouble(),
      windDirection: json['wind_direction']?.toDouble(),
      windGust: json['wind_gust']?.toDouble(),
      pressure: json['pressure']?.toDouble(),
      visibility: json['visibility']?.toDouble(),
      turbulence: json['turbulence']?.toDouble(),
      precipitation: json['precipitation']?.toDouble(),
      dataQuality: (json['data_quality'] ?? 1.0).toDouble(),
      weatherCondition: json['weather_condition'] as String?,
      extra: json['extra'] as Map<String, dynamic>?,
    );
  }
  final String source;
  final String? droneId;
  final DateTime timestamp;
  final double latitude;
  final double longitude;
  final double? altitude;
  final double? temperature;
  final double? humidity;
  final double? windSpeed;
  final double? windDirection;
  final double? windGust;
  final double? pressure;
  final double? visibility;
  final double? turbulence;
  final double? precipitation;
  final double dataQuality;
  final String? weatherCondition;
  final Map<String, dynamic>? extra;

  static DateTime _parseTimestamp(dynamic timestamp) {
    if (timestamp is int) {
      return DateTime.fromMillisecondsSinceEpoch(timestamp);
    }
    if (timestamp is String) {
      return DateTime.tryParse(timestamp) ?? DateTime.now();
    }
    return DateTime.now();
  }

  Map<String, dynamic> toJson() => {
        'source': source,
        'drone_id': droneId,
        'timestamp': timestamp.millisecondsSinceEpoch,
        'latitude': latitude,
        'longitude': longitude,
        'altitude': altitude,
        'temperature': temperature,
        'humidity': humidity,
        'wind_speed': windSpeed,
        'wind_direction': windDirection,
        'wind_gust': windGust,
        'pressure': pressure,
        'visibility': visibility,
        'turbulence': turbulence,
        'precipitation': precipitation,
        'data_quality': dataQuality,
        'weather_condition': weatherCondition,
        'extra': extra,
      };

  String get riskLevel {
    final ws = windSpeed;
    final vis = visibility;
    if (ws == null) return '未知';
    if (ws > 15 || vis != null && vis < 1) return '高';
    if (ws > 10 || vis != null && vis < 3) return '中';
    return '低';
  }
}
